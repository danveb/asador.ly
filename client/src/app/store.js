import { configureStore } from "@reduxjs/toolkit";
import authReducer from "../redux/auth/authSlice";
import pinReducer from "../redux/pins/pinSlice";

export const store = configureStore({
  reducer: {
    auth: authReducer,
    pins: pinReducer,
  },
}); 