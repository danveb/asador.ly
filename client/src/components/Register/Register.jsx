import { Link } from "react-router-dom"; 
import { Navbar, Menu } from "../index"; 
import "./Register.scss"; 

const Register = ({ menuOpen, setMenuOpen }) => {
    return (
        <div>
            <Navbar menuOpen={menuOpen} setMenuOpen={setMenuOpen} /> 
            <Menu menuOpen={menuOpen} setMenuOpen={setMenuOpen} /> 
            <div className="register">
                <div className="register-top">
                    <div className="register-title">
                        <h1>Register</h1>
                        <p>Create an account to share with us your favorite <span>parrillas</span> in Buenos Aires.</p>
                    </div>
                    <form className="register-form">
                        <label>Username</label>
                        <input 
                            type="text" 
                            placeholder="Create a username" 
                            required
                        />
                        <label>Email</label>
                        <input 
                            type="email" 
                            placeholder="Your email address" 
                            required
                        />
                        <label>Password</label>
                        <input 
                            type="password" 
                            placeholder="Your password" 
                            required
                        />
                        <button>Register</button>
                        <Link to="/login">Already have an account?</Link>
                    </form>
                </div>
            </div>
        </div>
    )
}

export default Register