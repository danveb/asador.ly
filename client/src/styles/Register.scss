.register {
  height: calc(100vh - 55px);
  width: 100%;
  padding: 40px 24px;
  display: flex;
  flex-direction: column;
  color: #000;
  background-color: var(--alice-blue);

  .register__wrapper {
    max-width: 1200px;
    margin: 0 auto;

    .register__head {

      h1 {
        font-size: 56px;
        font-family: "Prata", sans-serif;
        margin-bottom: 20px;

        @media screen and (max-width: 768px) {
          font-size: 46px;
        }

        @media screen and (max-width: 480px) {
          font-size: 36px;
        }
      }

      p {
        font-family: "Poppins", sans-serif;
        font-size: 15px;
        margin-bottom: 20px;

        @media screen and (max-width: 480px) {
          font-size: 13px;
        }

        span {
          font-style: italic;
        }
      }
    }

    .form {
      background-color: #fff;
      display: flex;
      flex-direction: column;
      justify-content: center;
      padding: 24px;
      border-radius: 10px;
      font-family: "Poppins", sans-serif;

      label {
        line-height: 1em;
        letter-spacing: .1em;
        font-size: 13px;
        text-transform: uppercase;
        color: #000;
        margin-bottom: 20px;
        font-family: inherit;
      }

      input {
        line-height: 1em;
        outline: 0;
        border: 1px solid #000;
        padding: 16px 22px;
        width: 100%;
        font-size: 12px;
        color: #000;
        margin-bottom: 20px;
        font-family: inherit;
      }

      button {
        height: 50px;
        width: 100%;
        background-color: #000;
        border: 2px solid #000;
        line-height: 1em;
        letter-spacing: .13em;
        font-size: 11px;
        font-family: "Arial", sans-serif;
        text-transform: uppercase;
        color: #fff;
        margin-bottom: 20px;
      }

      .login__link {
        display: flex;
        justify-content: flex-end;

        a {
          color: #3c3c3c;
          text-decoration: none;
          font-family: inherit;
          font-size: 14px;
          width: fit-content;
          cursor: pointer;
        }
      }

    }
  }
}