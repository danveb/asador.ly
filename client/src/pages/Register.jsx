import { useState, useEffect } from "react";
import { Link, useNavigate } from "react-router-dom";
import { Spinner } from "../components/index";
import { useSelector, useDispatch } from "react-redux";
import { register, reset } from "../redux/auth/authSlice";
import { toast } from "react-toastify";
import "../styles/Register.scss";

const Register = () => {
  const [formData, setFormData] = useState({
    username: "",
    email: "",
    password: "",
  });

  // destructure formData 
  const { username, email, password } = formData;

  // useNavigate 
  const navigate = useNavigate();

  // useDispatch
  const dispatch = useDispatch();

  const { user, isLoading, isError, isSuccess, message } = useSelector((state) => state.auth);

  // useEffect 
  useEffect(() => {
    if (isError) {
      toast.error(message, {
        position: "top-center",
        autoClose: 2000,
        pauseOnHover: true,
      });
    };
    if (isSuccess) {
      toast.success(`Welcome ${user.username} ðŸŽ‰ `, {
        position: "top-center",
        autoClose: 1500,
        pauseOnHover: false,
      });
      navigate("/map");
    };
    // redirect existing user back home
    if (user) {
      navigate("/");
    };
    dispatch(reset());
    // dependency   
  }, [user, isError, isSuccess, message, navigate, dispatch]);

  // handleChange
  const handleChange = (e) => {
    setFormData((prevState) => ({
      ...prevState,
      [e.target.name]: e.target.value
    }));
  };

  // handleSubmit 
  const handleSubmit = (e) => {
    e.preventDefault();
    const userData = {
      username,
      email,
      password
    };
    // error checking with toast... 
    if (username.length < 3) {
      toast.error("Please create a new username. Minimum length is 3 characters.")
    } else if (password.length < 4) {
      toast.error("Please create a new password. Minimum length is 4 characters.");
    } else {
      dispatch(register(userData));
    };
  };

  if (isLoading) {
    return <Spinner />
  };

  return (
    <div className="register">
      <div className="register-top">
        <div className="register-title">
          <h1>Register</h1>
          <p>Create an account to share with us your favorite <span>parrillas</span> in Buenos Aires.</p>
        </div>
        <form className="register-form" onSubmit={handleSubmit}>
          <label
            htmlFor="username"
          >Username</label>
          <input
            id="username"
            name="username"
            value={username}
            onChange={handleChange}
            type="text"
            placeholder="Create a username"
            required
          />
          <label
            htmlFor="email"
          >Email</label>
          <input
            id="email"
            name="email"
            value={email}
            onChange={handleChange}
            type="email"
            placeholder="Your email address"
            required
          />
          <label
            htmlFor="password"
          >Password</label>
          <input
            id="password"
            name="password"
            value={password}
            onChange={handleChange}
            type="password"
            placeholder="Your password"
            required
          />
          <button>Register</button>
          <Link to="/login">Already have an account?</Link>
        </form>
      </div>
    </div>
  )
}

export default Register