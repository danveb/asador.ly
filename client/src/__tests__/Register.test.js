import { render, fireEvent } from "@testing-library/react"; 
import '@testing-library/jest-dom/extend-expect';
import { store } from "../app/store"; 
import { Provider } from "react-redux";
import Register from "../components/Register/Register"; 
import { BrowserRouter } from "react-router-dom";

describe("Register component", () => {
    // smoke test
    test("renders Register component", () => {
        render(
            <Provider store={store}>
                <BrowserRouter>
                    <Register />
                </BrowserRouter>
            </Provider>
        );
    });

    // snapshot test
    test("matches snapshot", () => {
        const { asFragment } = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Register />
                </BrowserRouter>
            </Provider>
        );
        expect(asFragment()).toMatchSnapshot(); 
    });

    // getByText (with regex to find text)
    test("displays paragraph below header", () => {
        const { getByText } = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Register />
                </BrowserRouter>
            </Provider>
        );
        const paragraphTag = getByText(/Create an account to share with us your favorite/); 
        expect(paragraphTag).toBeInTheDocument(); 
    });

    // getAllByText
    test("displays Register text", () => {
        const { getAllByText } = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Register />
                </BrowserRouter>
            </Provider>
        );
        const title = getAllByText("Register"); 
        expect(title).toBeTruthy(); 
    });

    // getByLabelText
    test("displays username label", () => {
        const { getByLabelText } = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Register />
                </BrowserRouter>
            </Provider>
        );
        const usernameLabel = getByLabelText("Username"); 
        expect(usernameLabel).toBeInTheDocument(); 
    });

    // getByLabelText
    test("displays email label", () => {
        const { getByLabelText } = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Register />
                </BrowserRouter>
            </Provider>
        );
        const emailLabel = getByLabelText("Email"); 
        expect(emailLabel).toBeInTheDocument(); 
    });

    // getByLabelText
    test("displays password label", () => {
        const { getByLabelText } = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Register />
                </BrowserRouter>
            </Provider>
        );
        const passwordLabel = getByLabelText("Password"); 
        expect(passwordLabel).toBeInTheDocument(); 
    });

    // getByLabelText
    test("displays initial state for username/password", () => {
        const { getByLabelText } = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Register />
                </BrowserRouter>
            </Provider>
        );
        const usernameInput = getByLabelText("Username");
        const emailInput = getByLabelText("Email");
        const passwordInput = getByLabelText("Password");
        expect(usernameInput.value).toBe(""); 
        expect(emailInput.value).toBe(""); 
        expect(passwordInput.value).toBe(""); 
    });

    // getByLabelText, fireEvent
    test("displays username/email/password when added", () => {
        const { getByLabelText } = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Register />
                </BrowserRouter>
            </Provider>
        );
        const username = "jojo";
        const email = "jojo@gmail.com";
        const password = "jojo123"; 
        const usernameInput = getByLabelText("Username"); 
        const emailInput = getByLabelText("Email"); 
        const passwordInput = getByLabelText("Password"); 
        fireEvent.change(usernameInput, { target: { value: username }}); 
        fireEvent.change(emailInput, { target: { value: email }}); 
        fireEvent.change(passwordInput, { target: { value: password }}); 
        expect(usernameInput.value).toBe("jojo"); 
        expect(emailInput.value).toBe("jojo@gmail.com"); 
        expect(passwordInput.value).toBe("jojo123"); 
    });

    // getByRole 
    test("displays Register button", () => {
        const { getByRole } = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Register />
                </BrowserRouter>
            </Provider>
        );
        const registerBtn = getByRole("button", { name: "Register" }); 
        expect(registerBtn).toBeInTheDocument(); 
    });

    // mock form submission that receives username/password and submits
    test("mock form submission", () => {
        const username = "jojo";
        const email = "jojo@gmail.com";
        const password = "jojo123"; 
        const mockSubmit = jest.fn(); 
        const { getByLabelText, getByRole } = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Register onSubmit={mockSubmit(username, email, password)} />
                </BrowserRouter>
            </Provider>
        );
        const usernameInput = getByLabelText("Username"); 
        const emailInput = getByLabelText("Email"); 
        const passwordInput = getByLabelText("Password"); 
        const submitBtn = getByRole("button", { name: "Register" }); 
        fireEvent.change(usernameInput, { target: { value: username }}); 
        fireEvent.change(emailInput, { target: { value: email }}); 
        fireEvent.change(passwordInput, { target: { value: password }}); 
        fireEvent.click(submitBtn); 
        expect(mockSubmit).toHaveBeenCalled(); 
    });
});