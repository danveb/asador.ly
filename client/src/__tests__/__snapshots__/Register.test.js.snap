// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="navbar undefined"
    >
      <div
        class="navbar-left"
      >
        <a
          href="/"
        >
          Asador
        </a>
      </div>
      <div
        class="navbar-right"
      >
        <ul
          class="navbar-links"
        >
          <li>
            <a
              href="/"
            >
              Home
            </a>
          </li>
          <li>
            <a
              href="/about"
            >
              About
            </a>
          </li>
          <li>
            <a
              href="/login"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="AccountCircleIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
                />
              </svg>
            </a>
          </li>
        </ul>
        <div
          class="hamburger"
        >
          <span />
          <span />
          <span />
        </div>
      </div>
    </div>
    <div
      class="menu undefined"
    >
      <ul>
        <li>
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            href="/about"
          >
            About
          </a>
        </li>
        <li>
          <a
            href="/login"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AccountCircleIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
            </svg>
          </a>
        </li>
      </ul>
    </div>
    <div
      class="register"
    >
      <div
        class="register-top"
      >
        <div
          class="register-title"
        >
          <h1>
            Register
          </h1>
          <p>
            Create an account to share with us your favorite 
            <span>
              parrillas
            </span>
             in Buenos Aires.
          </p>
        </div>
        <form
          class="register-form"
        >
          <label
            for="username"
          >
            Username
          </label>
          <input
            id="username"
            name="username"
            placeholder="Create a username"
            required=""
            type="text"
            value=""
          />
          <label
            for="email"
          >
            Email
          </label>
          <input
            id="email"
            name="email"
            placeholder="Your email address"
            required=""
            type="email"
            value=""
          />
          <label
            for="password"
          >
            Password
          </label>
          <input
            id="password"
            name="password"
            placeholder="Your password"
            required=""
            type="password"
            value=""
          />
          <button>
            Register
          </button>
          <a
            href="/login"
          >
            Already have an account?
          </a>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
